peps/pep-0751.rst

```diff
@@ -0,0 +1,901 @@
+PEP: 751
+Title: A file format to list Python dependencies for installation reproducibility
+Author: Brett Cannon <brett@python.org>
+Status: Draft
+Type: Standards Track
+Topic: Packaging
+Created: 24-Jul-2024
+Replaces: 665
+
+Abstract
+========
+
+This PEP proposes a new file format for specifying what dependencies to
+(possibly) install into a Python environment for consistent installation
+repoducibility. The format is designed to be human-readable but
+machine-generated. Installers consuming the file should not need to resolve
+which listed dependencies to install based on other packages listed in the lock
+file, but instead evaluate each dependency in question in isolation.
+
+
+Motivation
+==========
+
+Currently, there is no standard way to specify what top-level dependencies one
+would like to see installed into a Python environment and then subsequently
+record what was eventually installed in a file (called a *lock file*).
+Considering there are four well-known solutions to this problem in the
+community (``pip freeze``, pip-tools_, Poetry_, and PDM_), there seems to be an
+appetite for lock files in general.
+
+Those tools also vary in what locking scenarios they support. For instance,
+``pip freeze`` and pip-tools only generate lock files for the current
+environment while PDM and Poetry try to lock for *any* environment to some
+degree. And none of them directly support locking to specific files to install
+which can be important for some workflows. There's also concerns around secure
+defaults in the face of supply chain attacks. Finally, not all the formats are
+easy to audit to determine what would be installed into an environment ahead of
+time.
+
+The lack of a standard also has some drawbacks. For instance, any tooling that
+wants to work with lock files must choose which format to support, potentially
+leaving users unsupported (e.g. if Dependabot_ chose not to support PDM,
+support by cloud providers who can do dependency installations on your behalf,
+etc.).
+
+
+Rationale
+=========
+
+The format is designed so that a *locker* which produces the lock file
+and an *installer* which consumes the lock file can be separate tools. This
+allows for situations such as cloud hosting providers to use their own installer
+that's optimized for their system which is independent of what locker the user
+used to create their lock file.
+
+The file format is designed to be human-readable. This is
+so that the contents of the file can be audited by a human to make sure no
+undesired dependencies end up being included in the lock file. It is also to
+facilitate easy understanding of what would be installed if the lock file
+without necessitating running a tool, once again to help with auditing. Finally,
+the format is designed so that viewing a diff of the file is easy by centralizing
+relevant details to minimizing needing to look at other parts of the file.
```

Comment:
```suggestion
without necessitating running a tool, once again to help with auditing. Finally,
the format is designed so that viewing a diff of the file is easy by centralizing
relevant details.
```